pkg/alertmanager/matcher_test.go

```diff
@@ -0,0 +1,161 @@
+package alertmanager
+
+import (
+	"testing"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/alertmanager/pkg/labels"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseMatcher(t *testing.T) {
+	tests := []struct {
+		name                      string
+		input                     string
+		expected                  *labels.Matcher
+		expectedTotal             float64
+		expectedDisagreeTotal     float64
+		expectedIncompatibleTotal float64
+		expectedInvalidTotal      float64
+		err                       string
+	}{{
+		name:          "is accepted",
+		input:         "foo=bar",
+		expected:      mustNewMatcher(t, labels.MatchEqual, "foo", "bar"),
+		expectedTotal: 1,
+	}, {
+		name:          "is regex accepted",
+		input:         "foo=~bar",
+		expected:      mustNewMatcher(t, labels.MatchRegexp, "foo", "bar"),
+		expectedTotal: 1,
+	}, {
+		name:                      "is accepted, but incompatible",
+		input:                     "foo=!bar\\n",
+		expected:                  mustNewMatcher(t, labels.MatchEqual, "foo", "!bar\n"),
+		expectedTotal:             1,
+		expectedIncompatibleTotal: 1,
+	}, {
+		name:                  "is accepted, but with disagreement",
+		input:                 "foo=\"\\xf0\\x9f\\x99\\x82\"",
```

Comment:
Disagreement happens here because `\xf0\x9f\x99\x82` is a 🙂 . This is not understood in the classic parser, just in the UTF-8 parser where it is interpreted as a 🙂 .